Fix #8797: Add OAuth2AuthenticationException to allowlist

napstr opened the pull request:
Add mixins for
- OAuth2AuthenticationException
- OAuth2Error

See #8797 for more details.

<!--
For Security Vulnerabilities, please use https://pivotal.io/security#reporting
-->

<!--
Before creating new features, we recommend creating an issue to discuss the feature. This ensures that everyone is on the same page before extensive work is done.

Thanks for contributing to Spring Security. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with gh-).
-->

napstr commented:
@jgrandja here's my take on #8797, what do you think?


Side note: I could not run the whole test suite locally to verify nothing else breaks. Master branch is failing locally with the same error as the CI build:
```
org.springframework.security.concurrent.CurrentDelegatingSecurityContextScheduledExecutorServiceTests > invokeAnyTimeout FAILED
    java.lang.NoSuchMethodError at PowerMockitoInjectingAnnotationEngine.java:40
```

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/jackson2/OAuth2AuthenticationExceptionMixin.java:
```diff
@@ -0,0 +1,43 @@
+ * @see OAuth2ClientJackson2Module
+ */
+@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS)
+@JsonIgnoreProperties(ignoreUnknown = true, value = {"cause", "stackTrace"})
```
Please add

```java
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY, getterVisibility = JsonAutoDetect.Visibility.NONE,
		isGetterVisibility = JsonAutoDetect.Visibility.NONE)
```

so it's consistent with the other mixins

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/jackson2/OAuth2ErrorMixin.java:
```diff
@@ -0,0 +1,44 @@
+ * @see OAuth2ClientJackson2Module
+ */
+@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS)
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Please add

```java
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY, getterVisibility = JsonAutoDetect.Visibility.NONE,
		isGetterVisibility = JsonAutoDetect.Visibility.NONE)
```

so it's consistent with the other mixins

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/jackson2/OAuth2ClientJackson2Module.java:
```diff
@@ -101,5 +103,7 @@ public void setupModule(SetupContext context) {
 		context.setMixInAnnotations(OidcUserAuthority.class, OidcUserAuthorityMixin.class);
 		context.setMixInAnnotations(DefaultOidcUser.class, DefaultOidcUserMixin.class);
 		context.setMixInAnnotations(OAuth2AuthenticationToken.class, OAuth2AuthenticationTokenMixin.class);
+		context.setMixInAnnotations(OAuth2AuthenticationException.class, OAuth2AuthenticationExceptionMixin.class);
```
Add `OAuth2AuthenticationExceptionMixin` and `OAuth2ErrorMixin` to class-level javadoc

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/jackson2/OAuth2AuthenticationExceptionMixinTests.java:
```diff
@@ -0,0 +1,88 @@
+	// @formatter:on
+
+	@Test
+	public void serializeOAuth2AuthenticationExceptionMixinTest() throws JsonProcessingException, JSONException {
```
Please rename test methods to following the naming convention **methodName**When**Condition**Then**Result** - see `OAuth2AuthorizedClientMixinTests` for examples

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/jackson2/OAuth2AuthenticationExceptionMixinTests.java:
```diff
@@ -0,0 +1,88 @@
+			+ "\n    \"uri\":null"
+			+ "\n  },"
+			+ "\n  \"message\":\"[authorization_request_not_found] \","
+			+ "\n  \"suppressed\":[\"[Ljava.lang.Throwable;\",[]],"
```
Add `suppressed` to `JsonIgnoreProperties`

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/jackson2/OAuth2AuthenticationExceptionMixinTests.java:
```diff
@@ -0,0 +1,88 @@
+
+	@Test
+	public void serializeOAuth2AuthenticationExceptionMixinTest() throws JsonProcessingException, JSONException {
+		OAuth2Error oauth2Error = new OAuth2Error("authorization_request_not_found");
```
Please add a `description` and `uri` to test this fully

napstr commented:
@jgrandja thanks for the detailed feedback. I think I have now addressed all your points, and expanded the test cases following the example set by other mixin tests.
Oh, also, using a local snapshot of this branch, I can confirm this solves the issue in my application.

jgrandja commented:
Thanks for the updates @napstr ! This is now in master and backported to 5.3.x
